hypothesis_regex.py

```diff
@@ -1,102 +1,268 @@
-    sre.CATEGORY_WORD: string.ascii_letters + string.digits + '_',
-    sre.CATEGORY_NOT_WORD: ''.join(
-        set(string.printable).difference(
-            string.ascii_letters + string.digits + '_'
-        )
-    )
-}
+HAS_SUBPATTERN_FLAGS = sys.version_info[:2] >= (3, 6)
+
+
+UNICODE_CATEGORIES = set([
+    'Cf', 'Cn', 'Co', 'LC', 'Ll', 'Lm', 'Lo', 'Lt', 'Lu',
+    'Mc', 'Me', 'Mn', 'Nd', 'Nl', 'No', 'Pc', 'Pd', 'Pe',
+    'Pf', 'Pi', 'Po', 'Ps', 'Sc', 'Sk', 'Sm', 'So', 'Zl',
+    'Zp', 'Zs',
+])
+
+
+SPACE_CHARS = u' \t\n\r\f\v'
+UNICODE_SPACE_CHARS = SPACE_CHARS + u'\x1c\x1d\x1e\x1f\x85'
+UNICODE_DIGIT_CATEGORIES = set(['Nd'])
+UNICODE_SPACE_CATEGORIES = set(['Zs', 'Zl', 'Zp'])
+UNICODE_LETTER_CATEGORIES = set(['LC', 'Ll', 'Lm', 'Lo', 'Lt', 'Lu'])
+UNICODE_WORD_CATEGORIES = UNICODE_LETTER_CATEGORIES | set(['Nd', 'Nl', 'No'])
+
+HAS_WEIRD_WORD_CHARS = (2, 7) <= sys.version_info[:2] < (3, 4)
+UNICODE_WEIRD_NONWORD_CHARS = u'\U00012432\U00012433\U00012456\U00012457'
+
+
+class Context(object):
+    __slots__ = ['groups', 'flags']
+
+    def __init__(self, groups=None, flags=0):
+        self.groups = groups or {}
+        self.flags = flags
+
+
+class CharactersBuilder(object):
+    def __init__(self, negate=False, flags=0):
+        self._categories = set()
+        self._whitelist_chars = set()
+        self._blacklist_chars = set()
+        self._negate = negate
+        self._ignorecase = flags & re.IGNORECASE
+        self._unicode = (not flags & re.ASCII) \
+            if six.PY3 else bool(flags & re.UNICODE)
+
+    @property
+    def strategy(self):
+        max_codepoint = None if self._unicode else 127
+
+        strategies = []
+        if self._negate:
+            if self._categories or self._whitelist_chars:
+                strategies.append(
+                    hs.characters(
+                        blacklist_categories=self._categories | set(['Cc', 'Cs']),
+                        blacklist_characters=self._whitelist_chars,
+                        max_codepoint=max_codepoint,
+                    )
+                )
+            if self._blacklist_chars:
+                strategies.append(
+                    hs.sampled_from(
+                        list(self._blacklist_chars - self._whitelist_chars)
+                    )
+                )
+        else:
+            if self._categories or self._blacklist_chars:
+                strategies.append(
+                    hs.characters(
+                        whitelist_categories=self._categories,
+                        blacklist_characters=self._blacklist_chars,
+                        max_codepoint=max_codepoint,
+                    )
+                )
+            if self._whitelist_chars:
+                strategies.append(
+                    hs.sampled_from(
+                        list(self._whitelist_chars - self._blacklist_chars)
+                    )
+                )
+
+        if len(strategies) == 0:
```

Comment:
Well, regular expression itself can't have an empty range (e.g. it can't be '[]'), but this is a separate builder object and might be used differently, so it should still maintain invariant: it returns a valid strategy that generates whatever you configured.